pkg/pquerier/collector.go

```diff
@@ -0,0 +1,184 @@
+	lastValuePerMetric := make(map[string]float64, len(ctx.columnsSpecByMetric))
+
+	for res := range responseChannel {
+		if res.IsRawQuery() {
+			rawCollector(ctx, res)
+		} else {
+			err := res.frame.addMetricIfNotExist(res.name, ctx.getResultBucketsSize(), res.IsServerAggregates())
+			if err != nil {
+				ctx.logger.Error("problem adding new metric '%v', lset: %v, err:%v", res.name, res.frame.lset, err)
+				ctx.errorChannel <- err
+				return
+			}
+
+			if res.IsServerAggregates() {
+				aggregateServerAggregates(ctx, res)
+			} else if res.IsClientAggregates() {
+				aggregateClientAggregates(ctx, res)
+			}
+
+			// It is possible to query an aggregate and down sample raw chunks in the same df.
+			if res.IsDownsample() {
+				lastTimePerMetric[res.name], lastValuePerMetric[res.name] = downsampleRawData(ctx, res, lastTimePerMetric[res.name], lastValuePerMetric[res.name])
+			}
+		}
+	}
+}
+
+func rawCollector(ctx *selectQueryContext, res *qryResults) {
+	frameIndex, ok := res.frame.columnByName[res.name]
+	if ok {
+		res.frame.rawColumns[frameIndex].(*V3ioRawSeries).AddChunks(res)
+	} else {
+		res.frame.rawColumns = append(res.frame.rawColumns, NewRawSeries(res, ctx.logger.GetChild("v3ioRawSeries")))
+		res.frame.columnByName[res.name] = len(res.frame.rawColumns) - 1
+	}
+}
+
+func aggregateClientAggregates(ctx *selectQueryContext, res *qryResults) {
+	it := newRawChunkIterator(res, nil)
+	for it.Next() {
+		t, v := it.At()
+		currentCell := (t - ctx.mint) / res.query.aggregationParams.Interval
+
+		for _, col := range res.frame.columns {
+			if col.GetColumnSpec().metric == res.name {
+				col.SetDataAt(int(currentCell), v)
+			}
+		}
+	}
+}
+
+func aggregateServerAggregates(ctx *selectQueryContext, res *qryResults) {
+	for _, col := range res.frame.columns {
+		if col.GetColumnSpec().metric == res.name &&
+			col.GetColumnSpec().function != 0 &&
+			col.GetColumnSpec().isConcrete() {
+
+			array, ok := res.fields[aggregate.ToAttrName(col.GetColumnSpec().function)]
+			if !ok {
+				ctx.logger.Warn("requested function %v was not found in response", col.GetColumnSpec().function)
+			} else {
+				// go over the byte array and convert each uint as we go to save memory allocation
+				bytes := array.([]byte)
+				for i := 16; i+8 <= len(bytes); i += 8 {
+					val := binary.LittleEndian.Uint64(bytes[i : i+8])
+					currentValueIndex := (i - 16) / 8
+					currentValueTime := res.query.partition.GetStartTime() + int64(currentValueIndex+1)*res.query.aggregationParams.GetRollupTime()
+					currentCell := (currentValueTime - ctx.mint) / res.query.aggregationParams.Interval
+
+					var floatVal float64
+					if aggregate.IsCountAggregate(col.GetColumnSpec().function) {
+						floatVal = float64(val)
+					} else {
+						floatVal = math.Float64frombits(val)
+					}
+					col.SetDataAt(int(currentCell), floatVal)
+				}
+			}
+		}
+	}
+}
+
+func downsampleRawData(ctx *selectQueryContext, res *qryResults,
```

Comment:
You can use the DataPoint struct instead of [`t`,`v`]
It was used for tests only so far, but feel free to move it into the library packages